### ISSUE TITLE
Generate Zelda Dungeons using Generative Grammars

### ISSUE THREAD

**Issue by jakeryang** [schrum2/MM-NEAT#341]
Using the technique outlined [here](http://www.t2thompson.com/wp-content/uploads/2016/03/zelda-aisb-2016.pdf) and more so [here](https://pdfs.semanticscholar.org/ab05/4224e6c43f96f06ece41f336b7a5a070a4e3.pdf) will allow us to specify dungeons based on a "sentence" describing what the player has to do to complete the dungeon. Then we can use that sentence and generate a level.

 To fully incorporate this we need to add new Zelda features per issue #330

**Comment by schrum2** [comment]
I was thinking specifically of using Grammatical Evolution, or even evolving the production rules. I think that is still the long term goal, but many steps need to be in place before we can evolve anything. First, we need to generate levels from a grammar. Here are some relevant papers.

https://www.brinckerhoff.org/tmp/grammatica_evolution_ieee_tec_2001.pdf
https://www.wi.uni-muenster.de/sites/wi/files/users/mpreu_02/games-material/ba-vm-ss2015/marco_niemann_-_constructive_generation_methods_for_dungeons_-_thesis.pdf
http://antoniosliapis.com/papers/multi-segment_evolution_of_dungeon_game_levels.pdf

**Comment by jakeryang** [comment]
The commit above is a starting point to generate a graph using grammars. So far, the overall idea is there but there are still some things that need to be worked out in the implementation. As of right now I'm getting a "Concurrent Modification Exception". I believe I understand why I'm getting that exception as I'm adding adjacencies to nodes while I'm still looping through the adjacencies. I will just need to figure out a way around it.

**Comment by jakeryang** [comment]
The method of replacing symbols with the grammars seem to work. Now to add a way of visualizing the graph.

**Comment by jakeryang** [comment]
The generated Zelda graphs now output to a .DOT file. Which the contents can be viewed here: http://sandbox.kidstrythisathome.com/erdos/ . However, the generated Zelda graphs are more like trees:

![chart](https://user-images.githubusercontent.com/6686189/59219099-4b332680-8b87-11e9-8b8c-d9db5b851eda.png)

The way the grammar is setup should prevent this, so I'll be working on this next

**Comment by jakeryang** [comment]
All of the adjacencies now show in the .DOT file resulting in a complete representation of the graph: 

![chart](https://user-images.githubusercontent.com/6686189/59224520-38731e80-8b94-11e9-982e-00a8d372c9b1.png)

The graph is still not very interesting, but the general implementation works, meaning we have to specify more grammar for there to be more of an exciting graph.

**Comment by jakeryang** [comment]
Latest commit boots up the rouge-like game with a generated dungeon, about a quarter of the time. However, there are some rooms that are in two different places and the generation process doesn't take into account future adjacencies

**Comment by jakeryang** [comment]
The issue of having multiple rooms with the same name is fixed and now I have an interesting result of a dungeon shaped as a heart. 
![ZeldaDungeonHeart](https://user-images.githubusercontent.com/6686189/59309893-66c32d80-8c6a-11e9-96b1-01c6fc6f7fcd.png)

This is the graph that was used:
![ZeldaGraphHeart](https://user-images.githubusercontent.com/6686189/59309957-907c5480-8c6a-11e9-8a6d-dc0a718a8d81.png)

**Comment by schrum2** [comment]
That dungeon doesn't seem to fully correspond to that graph (I think you alluded to this problem in your progress report). 

For example, near the top, two e nodes are connected by an edge, although they are not next to each other in the dungeon layout. There are also other discrepancies.

**Comment by jakeryang** [comment]
When looking at the papers, they seem to have the same problem of not having the dungeon fully correspond with the graph. There was more of an emphasis of making sure that the keys and locked doors were placed correctly to not have a dungeon that can't be complete.

**Comment by jakeryang** [comment]
More things have come up:
I tried to randomize the way the rooms were being placed and it ended up representing the graph more correctly, but it wouldn't make the necessary adjacencies
There are some dungeons were it would place two of the same room, again, but only on some graphs.

**Comment by jakeryang** [comment]
After a running the ZeldaGrammarGraph a few times and looking at the dungeon png, I noticed that the levels didn't go down as much as I thought, as there should be a random chance of placing a level going down, leading to little variation. I fixed this and now the levels are getting a lot more variation than before. However, some of the rooms aren't being placed such as they key and treasure rooms, but this happens a small percentage of the time.

**Comment by schrum2** [comment]
We're clearly generating dungeons using a grammar, though we're not evolving them yet. However, I think this issue can be closed. We might open a new issue for evolving things with the grammar, though random generation seems quite effective.
